dace/frontend/fortran/intrinsics.py

```diff
@@ -116,37 +178,166 @@ def replace_size(var: ast_internal_classes.Call_Expr_Node, scope_vars: ScopeVars
                                                      line_number=line)
 
-    def replace_real_kind(args: ast_internal_classes.Arg_List_Node, line):
-        if int(args.args[0].value) >= 9 or int(args.args[1].value) > 126:
+    def replace_real_kind(args: ast_internal_classes.Arg_List_Node, line, symbols: list):
+        if isinstance(args.args[0], ast_internal_classes.Int_Literal_Node):
+            arg0 = args.args[0].value
+        elif isinstance(args.args[0], ast_internal_classes.Name_Node):
+            if args.args[0].name in symbols:
+                arg0 = symbols[args.args[0].name].value
+            else:
+                raise ValueError("Only symbols can be names in selector")
+        else:
+            raise ValueError("Only literals or symbols can be arguments in selector")
+        if len(args.args) == 2:
+            if isinstance(args.args[1], ast_internal_classes.Int_Literal_Node):
+                arg1 = args.args[1].value
+            elif isinstance(args.args[1], ast_internal_classes.Name_Node):
+                if args.args[1].name in symbols:
+                    arg1 = symbols[args.args[1].name].value
+                else:
+                    raise ValueError("Only symbols can be names in selector")
+            else:
+                raise ValueError("Only literals or symbols can be arguments in selector")
+        else:
+            arg1 = 0
+        if int(arg0) >= 9 or int(arg1) > 126:
             return ast_internal_classes.Int_Literal_Node(value="8", line_number=line)
-        elif int(args.args[0].value) >= 3 or int(args.args[1].value) > 14:
+        elif int(arg0) >= 3 or int(arg1) > 14:
             return ast_internal_classes.Int_Literal_Node(value="4", line_number=line)
         else:
             return ast_internal_classes.Int_Literal_Node(value="2", line_number=line)
 
+    def replace_present(transformer: IntrinsicNodeTransformer, call: ast_internal_classes.Call_Expr_Node, line):
+
+        assert len(call.args) == 1
+        assert isinstance(call.args[0], ast_internal_classes.Name_Node)
+
+        var_name = call.args[0].name
+        test_var_name = f'__f2dace_OPTIONAL_{var_name}'
+
+        return (ast_internal_classes.Name_Node(name=test_var_name), "LOGICAL")
+
+    def replace_allocated(transformer: IntrinsicNodeTransformer, call: ast_internal_classes.Call_Expr_Node, line):
+
+        assert len(call.args) == 1
+        assert isinstance(call.args[0], ast_internal_classes.Name_Node)
+
+        var_name = call.args[0].name
+        test_var_name = f'__f2dace_ALLOCATED_{var_name}'
+
+        return (ast_internal_classes.Name_Node(name=test_var_name), "LOGICAL")
+
+    def replacement_epsilon(args: ast_internal_classes.Arg_List_Node, line, symbols: list):
+
+        # assert len(args) == 1
+        # assert isinstance(args[0], ast_internal_classes.Name_Node)
+
+        ret_val = sys.float_info.epsilon
```

Comment:
Epsilon value should depend on float type (if you use this, REAL4 will underflow)